### DISCUSSION TITLE
New ranking system?

### DISCUSSION THREAD

**Discussion by midnightPush** [cssbattle/forum#263] [Ideas]
I've been browsing the leaderboard and it's super depressing, sad and boring,
it's the same 3 persons in 99% of targets,
and they share code amongst each other (no proof, but it's 0% mathematical chance that such similarity it's an accident)

I think you should create something that more people from 200k accounts could enjoy, and not just 3, 🙄 
what's is the point of playing if you can't win? even from time to time

**Comment by chinchang** [comment]
Something on that front is getting cooked. 
On a side note, the top 3 don't have the same scores...so its not that they share code or something

**Reply by SelenIT** [reply]
Also, two of these top 3 came to CSSBattle relatively lately, so there is always a chance to break the old records :)

Sometimes the top scorers do share — not the exact code, but general ideas, approaches, and hints — but not just with each other, but also with other players, here and [in the Discord chat](https://discord.com/channels/909018468366549003/909018468936994848). Anyone can always ask them for a hint too!

**Reply by midnightPush** [reply]
So what are you guys waiting for?
More than 60% of targets have very little to do with CSS itself. The trend is clear, and it's not optimistic. In the last battle, only one target was solved without unicode and clip-path cheat codes .
Does it need to be 100% for a wake-up call?
When I talk to people, they are usually unhappy about unicode solutions, since it's a totally useless skill and a waste of time, but that just allows for higher scores.

**Reply by gunnarbittersmann** [reply]
What’s your solution? Have you even thought about if there is one?

`clip-path`: It’s a CSS property; `clip-path: inset()` should be allowed. You want to forbid the special value with `path()`? Do you want to forbid the `path()` function also with other properties like `offset`, or just with `clip-path`? The makers of the game tried the latter – and failed. The combination can still be used using another CSS feature. (Is a CSS feature a hack?) They might try to not accept such solutions – and will fail again. There will always be a way to use the combination of `clip-path` and `path()`.

Unicode: You want to forbid characters like `●` U+25CF and `·` (U+00B7)? Then you must also forbid `.` (U+002E). Since that’s also the decimal point, you must not forbid that character generally, but only as text content. This would lead to the conclusion that no text content at all would be allowed, including white space characters. Now that would annoy many people when they cannot format their code using line breaks and indentation. Removing any text content from the code on the server before calculating the score is no option either, because things might render diffently without white space, which again would annoy a lot of people.

As I see it, there’s only one way: You have to deal with `clip-path: path()` and with Unicode.

**Reply by SelenIT** [reply]
There could be a potential way to prevent overusing of Unicode/text for top solutions by using two backend renderers with different default fonts, and counting the solution as 100% match only if both renderers get 100%. But this would mean double load on the backend and possible performance impact.

Regarding `clip-path:path()`, I agree that it's a standard CSS feature and shouldn't be treated any different from `offset:path()`. It also seems to have clear superiority in one specific kind of targets — single-color symmetrical shapes, like [#56](https://cssbattle.dev/play/56), [#61](https://cssbattle.dev/play/61), or [#66](https://cssbattle.dev/play/66). So I guess that if new targets will be more colorful and have more non-symmetrical details then this feature won't give much advantage and there would be not much reason to restrict it.

**Reply by midnightPush** [reply]
> As I see it

You only see what you believe, which at the moment is learned helplessness - "nothing can be done/improved"

**Reply by gunnarbittersmann** [reply]
I have provided arguments for my conclusion. Prove me wrong with counter-arguments. With idle talk, you won’t.

**Reply by LukasBohmRosenbauer** [reply]
@chinchang man, are you still cooking? why does it take so long, the ranking is already rotting and stinks badly,
current version has absolutely no sense and meaning as those 4s:
![LDE9DpV](https://user-images.githubusercontent.com/119531802/204901352-ada57d1d-6bd5-4343-8873-0124f512e8ed.png)

clear shares (cough... cheaters) preserve in buddies contest (not sure about the rest in top10), ruin the fun for rest 200k people, that is highly unfair and unsportsmanlike,

how anybody can take it seriously and even pay for it? if cronyism gets on top and wins
sounds ridiculous to me

**Reply by gunnarbittersmann** [reply]
Just no. No sharing. We already have discussed why some players have top solutions of the same-length. You might want to start reading at https://github.com/cssbattle/forum/discussions/263#discussioncomment-3291254.

**Reply by LukasBohmRosenbauer** [reply]
lol, it's obvious, no need for excuses, everyone can see leaderboard, if one of them make a score the rest will follow automatically like ants with the same or highly similar score in a short amount of span, it's not a random accdient

btw santa claus is also not real

**Reply by gunnarbittersmann** [reply]
@SelenIT wrote:
> There could be a potential way to prevent overusing of Unicode/text for top solutions by using two backend renderers with different default fonts, and counting the solution as 100% match only if both renderers get 100%.

I don’t think this would help. Fonts don’t have glyphs for the whole Unicode range. If the browser doesn’t find a glyph for a given character in the current font, it takes a glyph from a fallback font. With characters like ▲, ◼ or ●, there might be only one symbol fonts with glyphs for them, i.e. all renderers would fall back to the very same font and return the same result.

**Reply by gunnarbittersmann** [reply]
> lol, it's obvious, no need for excuses, everyone can see leaderboard, if one of them make a score the rest will follow automatically like ants with the same or highly similar score in a short amount of span, it's not a random accdient

I’ve just pointed you to an [explanation](https://github.com/cssbattle/forum/discussions/263#discussioncomment-3291254) why this happens. You might want to read it.

**Comment by SelenIT** [comment]
Maybe it would be possible to make separate leaderboards for different methods of solution, in order to enable independent competition for each column of [this table](https://docs.google.com/spreadsheets/d/18pO9ESpPTP7w6ERWNPHNSI03CQaSS5S_42nHSo_imbw/edit#gid=0)?

**Reply by gunnarbittersmann** [reply]
Where would solutions that use both Unicode and `clip-path` fit in?

**Reply by Beowolve** [reply]
On this table the rule is simple.
If **clip-path:path** is used it is counted in that row with or without unicode.

**Comment by rafehimself** [comment]
@chinchang 
do we need to increase medals pool in case of same scores? I find it highly unsatisfying that there is no reward for finding best solution, because top 3 spots are usually already besieged.

![image](https://user-images.githubusercontent.com/58576132/177800217-fddf6c77-beeb-4351-bd5e-c30f7e406366.png)

the code itself it's not worse in any sense, it's just matter of timing ⏲️ ,which we know from over a 100 years that's it's relative, not absolute, that should be awarded on the profile at least,
althought I'm not sure how to make that work in open targets ..., but you will figure that out ;)

please take that into account, if there are changes in coming

**Reply by gunnarbittersmann** [reply]
> because top 3 spots are usually already besieged.

Just like https://cssbattle.dev/play/40:

1. [Ilya Streltsyn](https://cssbattle.dev/player/selenit) 741.54 (109 Chars)
2. [Beo](https://cssbattle.dev/player/beo) 741.54 (109 Chars)
3. [Jeroen Meijers](https://cssbattle.dev/player/jeme) 741.54 (109 Chars)

Oh, that’s yesterday’s news. 😆

---

> it's just matter of timing ⏲️ 

Look who’s talking! Mister “matter of timing” himself, of whose gold most medals most just come just from being the first one to rewrite transform functions into `rotate` and `scale` properties once the server has accepted them. And some of whose silver medals just come being the first one to copy the top solutions on open targets,

**Reply by rafehimself** [reply]
[try this](https://www.ministerstwogadzetow.com/media/catalog/product/cache/1/image/1800x/040ec09b1e35df139433887a97daa66f/m/a/masc_na_bol_dupy_2.jpg) 🤡

**Comment by Beowolve** [comment]
Many sports solve this by giving people with same score the same rank.

**For example target #116:**
![image](https://user-images.githubusercontent.com/32326640/177808038-901b22f8-5ff4-4c07-ac3b-bc3afa973dac.png)

**Would then change to:**

1 - [Jeroen Meijers](https://cssbattle.dev/player/jeme)
1 - [Beo](https://cssbattle.dev/player/beo)
1 - [Rafe](https://cssbattle.dev/player/rafe)
2 -  [Gabriele](https://cssbattle.dev/player/gabriele19)
3 - [Vincent De Oliveira](https://cssbattle.dev/player/iamvdo)
4 - [Ilya Streltsyn](https://cssbattle.dev/player/selenit)
5 - [Janne Ala-Äijälä](https://cssbattle.dev/player/underdude)
5 - [Gunnar Bittersmann](https://cssbattle.dev/player/gunnar)
6 - [出雲正尚](https://cssbattle.dev/player/3XJMTE0a3AaXoB0VoKuru7rQU362)
7 - [Mr G](https://cssbattle.dev/player/razvanm)

**Reply by gunnarbittersmann** [reply]
No. It would change to:

1 - [Jeroen Meijers](https://cssbattle.dev/player/jeme)
1 - [Beo](https://cssbattle.dev/player/beo)
1 - [Rafe](https://cssbattle.dev/player/rafe)
4 - [Gabriele](https://cssbattle.dev/player/gabriele19)
5 - [Vincent De Oliveira](https://cssbattle.dev/player/iamvdo)
6 - [Ilya Streltsyn](https://cssbattle.dev/player/selenit)
7 - [Janne Ala-Äijälä](https://cssbattle.dev/player/underdude)
7 - [Gunnar Bittersmann](https://cssbattle.dev/player/gunnar)
9 - [出雲正尚](https://cssbattle.dev/player/3XJMTE0a3AaXoB0VoKuru7rQU362)
10 - [Mr G](https://cssbattle.dev/player/razvanm)

The 4th wouldn’t get silver just because there are currently 3 gold medalists.

**Reply by joewinbpaul** [reply]
But then if that is the case..... WHAT ABOUT TARGET 1???!!! A lot of people got 54 chars (including me) but didn't get First Rank. Even the guy in second place got a silver medal, not gold.

**Reply by joewinbpaul** [reply]
I don't think it would change

**Comment by gunnarbittersmann** [comment]
> I've been browsing the leaderboard and it's super depressing, sad and boring, it's the same 3 persons in 99% of targets, and they share code amongst each other (no proof, but it's 0% mathematical chance that such similarity it's an accident)

There’s a simple reason that sometimes multiple players have the same result: When you reach the current top result, you tend to think that it doesn’t get any better, and you stop trying to improve further. Until somebody breaks the record – then it starts again. And eventually you find a solution that you weren’t looking for before you were aware that it is possible.

And the same score doesn’t mean the same solution. Take [#40](https://cssbattle.dev/play/40), for example. When @iamvdo also reached 108, I asked him about a detail in his solution, and from his answer I came to the conclusion that our solutions must be very different. I guess we are even using different techniques. The only thing our solutions have in common might be the color values – and accidentally the code length.

TL;DR: You were wrong with “0% mathematical chance”.

**Reply by midnightPush** [reply]
No, you're just naive, although that might be the case sometimes, it's not true for a big picture
You make conculsion from 1 example instead of looking at 100+ pattern,
funny that Gabriele loved it, while being the biggest beggar and sharer of the code of others🤣

**Reply by JDMeijers** [reply]
I don't think you are seeing the big picture here @midnightPush. As @gunnarbittersmann mentioned, when playing for the top solution, you get triggered for finding last characters. By playing here for a long time, and knowing a lot of techniques, top players will be close to each other most of the time. On bigger targets like #108, there is more room for different approaches, techniques etc.. But smaller targets like #109, you get pushed into the same technique. And because of knowing a lot of techniques and tricks, top players know where and how to find the last characters.

And as Gunnar also mentioned, he used #40 as an example. I can write an example for almost every target. So "Make conclusion from 1 example instead of looking at 100+ pattern" isn't true.

**Reply by Gabb192** [reply]
@midnightPush You are probably the naive here, accusing me of sharing someone else's code is really pathetic. Feel free to ask if any of the top players shared a TOP solution to me (be careful you are not talking about an idea) or if even after they shared a top solution I went to share it with someone else. I do not deny that when I was not yet in the top 15 of the general rank I exchanged **MY** code with someone else's code, I emphasize **mine**. If you think that we exchange solutions in the top rankings, I am sorry but you are wrong, but obviously **personal** ideas and attempts with more characters I do not deny that sometimes it is done. I don't know you as your name is not in the global top 100, you probably use a different name on GitHub, but I have an idea who you might be, you are probably just envying the fact that I am above you in the rankings. 😉

**Reply by joewinbpaul** [reply]
> I don't think you are seeing the big picture here @midnightPush. As @gunnarbittersmann mentioned, when playing for the top solution, you get triggered for finding last characters. By playing here for a long time, and knowing a lot of techniques, top players will be close to each other most of the time. On bigger targets like #108, there is more room for different approaches, techniques etc.. But smaller targets like #109, you get pushed into the same technique. And because of knowing a lot of techniques and tricks, top players know where and how to find the last characters.
> 
> And as Gunnar also mentioned, he used #40 as an example. I can write an example for almost every target. So "Make conclusion from 1 example instead of looking at 100+ pattern" isn't true.

umm... can u give a hint on what technique u guys used in target #109 to get 77 chars? I used unicode (even tho I'm a beginner in unicode) and got 116 chars - which is my personal best..

**Reply by chinchang** [reply]
To end the discussion here - "top solutions are not the same even though they are of the same length". You have to trust us on this because we do random checks for top solutions. What you see is "same score" and conclude that solution must be the same. But "same score" comes from same "code length", not "same solution".

**Reply by LukasBohmRosenbauer** [reply]
im might not know all the scores, but I know enough to say what you're saying is simply not true, 
do you know titanic story right?
good luck with keeping your delusions alive instead of facing facts

**Reply by PatriciaLincoln** [reply]
@chinchang 
many times it feels like you lack the "feeling" of the community, a player perspective and their needs, totally blind for their feedback 🙈 

have you ever wonder to give a away marketing and management to someone more competent? because that might lead to bottle neck in the project, and just focusing on your strengths, whatever they are coding, developing etc

that's just my perspective, don't take it personally, nobody is great at everything 🙄

**Reply by SelenIT** [reply]
> im might not know all the scores…

Well, let's consider [Target #130](https://cssbattle.dev/play/130). It's an open Target, so we can compare submissions by different people.

All 4 top solutions have the same length and are *conceptually* the same. They all have the `bgcolor` attribute before `style` and the `filter` property at the end because it allows to omit the trailing parenthese and thus save 1 character. But other properties have different order. Different case of the color value. Slightly different numbers in color values (a rather common trick to overcome the 99.9% problem).

What do you believe is more likely: that all these people edited each other's solutions to make them look different (while others just copied the same top solution) or that they independently came to the conclusion that `filter:drop-shadow` is the best for this target?

**Reply by GalaxyBrain100** [reply]
1. Cherry-picking examples to support personal beliefs doesn't prove anything, just like one sunny day in England doesn't change the fact it's a rainy climate country, this kind of irrational reasoning is pointless, should not be part of aware adults, at best it can be used to create mythology, so please stop it, or maybe we want to create cssbattle religion?😆 

@chingchang
2. We didn't ask you to investigate whether it's true or not, because we KNOW this is HAPPENING from first-hand experiences, so do not try to convince us that our experience is invalid ... gosh 🥶  learn to listen instead? 🤔 
Your answer is even more concerning due to the fact showing your totally unawareness of how the scores are made and how the game is working in practice (and it's not first time, clip-path secret 😉 ), sharing, trading, passing information for favors, working in groups, asking for hints/tips or just copy/paste is common practice, regardless of those few deniers are saying,  the ranking flipped side down from the moment of introducing discord. (please don't try to create another myths that it didn't change anything)
 
The reason of the thread was to make you more aware: 
a) this is how the system currently is work
b) it has no sense whatsoever
c) can you come up with a better solution?

Maintaining that kind of ranking is only hypocrisy, nothing more
Have a nice day.

**Reply by Beowolve** [reply]
There is a strong competitive mindset for all top players. If you really think I give away my top solutions where I worked for hours, days or sometimes weeks to get that score just to get beaten by someone else who manages to find a char I missed, then you are just naive or a troll.

If you need help/tips, come and ask in discord and you will get enough tips to be able to find it yourself. I would have not come to the top of the leaderboard without learning from other players. But I find my solutions myself, like all the other top players do. If anyone of us makes a new record everyone wants to find it. Most of the time you get a good clue from the records table. Most players are pretty open of the method used and at least tell if it is unicode/css/clip and you will find that information there.

Everyone complaining here, that they are not able to get a top score/win are just not accepting the fact that it takes a good amount of commitment just for a single target to try a lot of versions/submissions to make top scores. I have many thousand submissions on some targets and sometimes all of this just to find one single char less than the competition.

So basically, you have no clue what you are talking about...

**Reply by SelenIT** [reply]
@GalaxyBrain100 ok, so you prefer scientific method. Let's make an experiment then. You pick any 3 targets where I have 🥇 (so you can't accuse me of cherry-picking). I give you all the selectors, all the CSS properties (including custom ones), and all the CSS units I used in my top solutions there (it's way, way more info than the hints that top players occasionaly give to each other!).

If this info would be enough for you to get the maximum currently available score in at least two of these 3 targets, I will admit your point that the scoring system is broken. Otherwise, you will admit that sharing such occasional hints is not the main reason why the current top scorers got their "privilege", apologize to @chinchang, and help to prove @LukasBohmRosenbauer wrong as well.

Are you ready for this challenge? 😉

**Reply by LukasBohmRosenbauer** [reply]
yeah looks like breaking #101 at same time is another magical accident (one of hundreds cases) we should take on faith from his majesty and his minions, seems legit to me 🤔 😆

**Reply by SelenIT** [reply]
@LukasBohmRosenbauer every time somebody breaks the record (and thus proves that shorter solution exists) it motivates others to resume experimenting on that target. Sometimes just resubmitting previous solution with 99.9% match works after the renderer update (like it was with [target #3](https://cssbattle.dev/play/3)), sometimes making some more (dozens of) attempts finally pays off where a player previously gave up. No magic, no secret conspiracy, just human psychology and competition:)

BTW, you can accept my challenge above and prove me wrong as well 😉

**Reply by Gabb192** [reply]
@LukasBohmRosenbauer So according to your reasoning I shared the first record that I found on https://github.com/cssbattle/forum/discussions/101 to the other top players so they can find a better (or at least the same) solution and beat me? What's the point of doing that? Losing the first place and hours of trying?

I really don't understand what you guys are trying to prove.
